sensor: add TI FDC2X1X sensor driver and sample application

skullbox305 opened the pull request:
This patch adds support for the Texas Instruments FDC2X1X Capacitance-to-Digital Converter for Proximity and Level Sensing Applications. 
Along with the driver, a sample sensor application is provided which tests the device driver.

Signed-off-by: Igor Knippenberg <igor.knippenberg@gmail.com>

mbolivar-nordic commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -0,0 +1,1010 @@
+				      FDC2X1X_TO_I2C_REG(reg),
+				      (uint8_t *) data, length);
+	} else {
+		if (length <= 0) {
```
What's going on here? The `length` argument is a `size_t`, which is unsigned, so this doesn't look right.

Should this be something like `if (length != 2)` instead? The `memcpy(buf + 1, data, sizeof(uint16_t));` line below makes it seem like exactly two bytes of `data` are being used in this case.

skullbox305 replied:
your are right, my bad

mbolivar-nordic commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -0,0 +1,1010 @@
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int fdc2x1x_bus_access(const struct device *dev, uint8_t reg,
+			      void *data, size_t length)
```
Why not just declare this `uint8_t *data` and avoid casting it to `uint8_t*` below? The only two callers are allocating uint8_t arrays for this argument.

skullbox305 replied:
declared as `uint8_t *data` now

mbolivar-nordic commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -0,0 +1,1010 @@
+ * @param enable - True = enable shutdown, false = disable shutdown
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int fdc2x1x_set_shutdown(const struct device *dev, bool enable)
```
This should be `static`.

mbolivar-nordic commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -0,0 +1,1010 @@
+		new_state = *(uint32_t *)context;
+
+		if (new_state != data->pm_state) {
+
```
Nit: unnecessary whitespace.

mbolivar-nordic replied:
> Nit: unnecessary whitespace.

Still here.

skullbox305 replied:
sorry, my bad. Removed the whitespace now

mbolivar-nordic commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -0,0 +1,1010 @@
+ * Set attributes for the device.
+ * @param dev - The device structure.
+ * @param chan - The sensor channel type.
+ * @param attr - The sensor atrribute.
+ * @param value - The sensor atrribute value.
```
Nit: "attribute" typos

mbolivar-nordic commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -0,0 +1,1010 @@
+			return ret;
+		}
+
+		ret =
+			fdc2x1x_set_fref_divider(dev, ch,
```
Nit: this looks like an unnecessary extra newline.

mbolivar-nordic commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -0,0 +1,1010 @@
+	gpio_pin_configure(data->gpio, cfg->sd_pin,
+			   GPIO_OUTPUT_INACTIVE | cfg->sd_flags);
+
+	gpio_pin_set(data->gpio, cfg->sd_pin, 0);
```
This seems redundant with the GPIO_OUTPUT_INACTIVE flag above, no?

skullbox305 replied:
yes this call is unecessary. I removed it

mbolivar-nordic commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -0,0 +1,1010 @@
+		ch_supported = 2;
+	}
+
+	if ((cfg->num_channels <= 0)) {
```
Nit: extraneous parentheses

mbolivar-nordic commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -0,0 +1,1010 @@
+		return -EINVAL;
+	}
+
+	k_sleep(K_MSEC(100));
```
Why? This is going to block the init sequence for 100 msec.

skullbox305 replied:
Forgot to remove it after testing. Removed it now

mbolivar-nordic commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -0,0 +1,1010 @@
+	.sd_flags = DT_INST_GPIO_FLAGS(n, sd_gpios), \
+
+#define FDC2X1X_SD(n)					\
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(n, sd_gpios),	\
+		    (FDC2X1X_SD_PROPS(n)),		\
+		    ())
```
Feel free to leave this as-is, but this could be written more simply as:

```
	IF_ENABLED(DT_INST_NODE_HAS_PROP(n, sd_gpios),	\
		    (FDC2X1X_SD_PROPS(n)))
```

skullbox305 replied:
I took your suggestions. It is definitely easier to read

mbolivar-nordic commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -0,0 +1,1010 @@
+	.intb_flags = DT_INST_GPIO_FLAGS(n, intb_gpios), \
+
+#define FDC2X1X_INTB(n)			     \
+	COND_CODE_1(CONFIG_FDC2X1X_TRIGGER,  \
+		    (FDC2X1X_INTB_PROPS(n)), \
+		    ())
```
Same here. Feel free to leave as-is, but this could use IF_ENABLED also.

mbolivar-nordic commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -0,0 +1,1010 @@
+		    (FDC2X1X_INTB_PROPS(n)), \
+		    ())
+
+#define FDC2X1X_CH_CFG_INIT(m)				  \
```
I think `ch` would be a better variable name than `m`. Here, `m` after `n` suggests that `m` is also an instance number, but it's actually a node identifier.

skullbox305 replied:
changed the variable name to `ch`

mbolivar-nordic commented on dts/bindings/sensor/ti,fdc2x1x.yaml:
```diff
@@ -0,0 +1,235 @@
+      description: |
+        Set to identify the sensor as FDC2114 or FDC2214 (4-channel version)
+
+    autoscan:
```
Couldn't this just be boolean?

skullbox305 replied:
Yes a boolean makes more sense. I changed that

mbolivar-nordic commented on dts/bindings/sensor/ti,fdc2x1x.yaml:
```diff
@@ -0,0 +1,235 @@
+    rr-sequence:
+      type: int
+      required: false
+      default: 0
```
Defaults in bindings must be justified; please explain why its value is set in the description.

If this value corresponds to the power-on-reset behavior of the device and the property can be changed independently of other properties, that's OK. If this is an individual preference made to work with the sample application or something like that, the property should not have the default value (and may need to be made `required` instead if there's no sane default corresponding to PoR behavior or something like that).

skullbox305 replied:
Most of those values are power-on-reset defaults. The ones which are not I changed to required. I also added an explanation in each description for this and your comments below

mbolivar-nordic commented on dts/bindings/sensor/ti,fdc2x1x.yaml:
```diff
@@ -0,0 +1,235 @@
+    active-channel:
+      type: int
+      required: false
+      default: 0
```
Same comments about default made above apply here also.

mbolivar-nordic commented on dts/bindings/sensor/ti,fdc2x1x.yaml:
```diff
@@ -0,0 +1,235 @@
+    deglitch:
+      type: int
+      required: false
+      default: 5
```
Same comments about default made above apply here also.

mbolivar-nordic commented on dts/bindings/sensor/ti,fdc2x1x.yaml:
```diff
@@ -0,0 +1,235 @@
+    sensor-activate-sel:
+      type: string
+      required: false
+      default: "low-power"
```
Same comments about default made above apply here also.

mbolivar-nordic commented on dts/bindings/sensor/ti,fdc2x1x.yaml:
```diff
@@ -0,0 +1,235 @@
+    ref-clk-src:
+      type: string
+      required: false
+      default: "internal"
```
Same comments about default made above apply here also.

mbolivar-nordic commented on dts/bindings/sensor/ti,fdc2x1x.yaml:
```diff
@@ -0,0 +1,235 @@
+    current-drive:
+      type: string
+      required: false
+      default: "normal"
```
Same comments about default made above apply here also.

mbolivar-nordic commented on dts/bindings/sensor/ti,fdc2x1x.yaml:
```diff
@@ -0,0 +1,235 @@
+    output-gain:
+      type: int
+      required: false
+      default: 0
```
Same comments about default made above apply here also.

mbolivar-nordic commented on dts/bindings/sensor/ti,fdc2x1x.yaml:
```diff
@@ -0,0 +1,235 @@
+      rcount:
+        type: int
+        required: false
+        default: 7499
```
Same comments about default made above apply here also.

mbolivar-nordic commented on dts/bindings/sensor/ti,fdc2x1x.yaml:
```diff
@@ -0,0 +1,235 @@
+      offset:
+        type: int
+        required: false
+        default: 0
```
Same comments about default made above apply here also.

skullbox305 replied:
This value applies only to the 12 Bit version. So I dont think it makes sense to make `required`. I leave the default power-on-reset value as `default`

mbolivar-nordic commented on dts/bindings/sensor/ti,fdc2x1x.yaml:
```diff
@@ -0,0 +1,235 @@
+      settlecount:
+        type: int
+        required: false
+        default: 48
```
Same comments about default made above apply here also.

mbolivar-nordic commented on dts/bindings/sensor/ti,fdc2x1x.yaml:
```diff
@@ -0,0 +1,235 @@
+      fref-divider:
+        type: int
+        required: false
+        default: 1
```
Same comments about default made above apply here also.

mbolivar-nordic commented on dts/bindings/sensor/ti,fdc2x1x.yaml:
```diff
@@ -0,0 +1,235 @@
+      idrive:
+        type: int
+        required: false
+        default: 10
```
Same comments about default made above apply here also.

mbolivar-nordic replied:
> Same comments about default made above apply here also.

This `default` still does not have a justification in the description. The other ones look good now; thanks.

skullbox305 replied:
Sorry, forgot to remove the default

mbolivar-nordic commented on dts/bindings/sensor/ti,fdc2x1x.yaml:
```diff
@@ -0,0 +1,235 @@
+      fin-sel:
+        type: int
+        required: false
+        default: 2
```
Same comments about default made above apply here also.

mbolivar-nordic commented on include/drivers/sensor.h:
```diff
@@ -174,6 +174,15 @@ enum sensor_channel {
 	/** Desired charging current in mA */
 	SENSOR_CHAN_GAUGE_DESIRED_CHARGING_CURRENT,
 
+	/** Channel 0, raw data */
+	SENSOR_CHAN_RAW_CH0,
+	/** Channel 1, raw data */
+	SENSOR_CHAN_RAW_CH1,
+	/** Channel 2, raw data */
+	SENSOR_CHAN_RAW_CH2,
+	/** Channel 3, raw data */
+	SENSOR_CHAN_RAW_CH3,
+
```
I'm not sure this makes sense to add here. @MaureenHelm ?

skullbox305 replied:
I was also not happy about that, but I don't think the current sensor API is able to read multiple channels of the same type from a single device? Would be good to be able to pass something like a channel number to the channel get function e.g. `sensor_channel_get(dev, SENSOR_CHAN_RAW, ch_number, buffer);`. Is there a plan to add a functionality like that?

mbolivar-nordic replied:
This is a decision for the sensor maintainers; I can't comment on this one.

MaureenHelm replied:
I'm not a fan of adding channels without a type or standard unit. What does the quantity represent?

skullbox305 replied:
It can be represented as Capacitance or Frequency. I can add those conversion. Should I leave the Channel 0-3  numbering at the end of each type?

skullbox305 replied:
Added the conversions from raw to Capacitance and Freqzency.

mbolivar-nordic commented on samples/sensor/fdc2x1x/README.rst:
```diff
@@ -0,0 +1,79 @@
+********************
+
+This sample outputs sensor data to the console and can be read by any serial
+console program. It should work with any platform featuring a I2C interface.
```
You should add a note that different platforms will require custom devicetree overlays.

skullbox305 replied:
Added your suggested note

mbolivar-nordic commented on samples/sensor/fdc2x1x/boards/nrf9160dk_nrf9160.overlay:
```diff
@@ -0,0 +1,22 @@
+		sd-gpios = <&gpio0 16 GPIO_ACTIVE_HIGH>;
+		intb-gpios = <&gpio0 20 GPIO_ACTIVE_LOW>;
+
+		// using the default channel settings
```
Judging by this comment, I think the binding should probably not have default settings, but instead the properties should be made `required` and this sample can provide some sensible defaults for evaluation here.

skullbox305 replied:
Made all channel bindings, except for `offset`, required

mbolivar-nordic commented on samples/sensor/fdc2x1x/src/main.c:
```diff
@@ -0,0 +1,134 @@
+		device_get_binding(DT_LABEL(DT_INST(0, ti_fdc2x1x)));
+
+	if (dev == NULL) {
+		printk("Could not get %s device \n",
```
Extra whitespace here.

mbolivar-nordic commented on samples/sensor/fdc2x1x/src/main.c:
```diff
@@ -0,0 +1,134 @@
+void main(void)
+{
+	struct sensor_value ch_buf[] = {
+		DT_FOREACH_CHILD(DT_INST(0, ti_fdc2x1x), CH_BUF_INIT)
```
@galak @pabigot I'm curious about what you think about the continued use of DT_INST() in sample applications.

I'm not really in favor of it because it means that users have to understand instance numbers, which are complicated and basically just for driver authors. Getting instance numbers to work in a sane way outside of a driver breaks down fast when there are multiple devices of the same compatible.

skullbox305 replied:
I wanted to make the sample application work with both, the 2 channel and the 4 channel model of the sensor. I didn't know how else to get this information inside the sample. Should I instead make it only read 2 channels for both models?

mbolivar-nordic replied:
Let's keep it for now and revisit if we come up with a coherent policy in this regard.

mbolivar-nordic reviewed (changes requested):
Thanks for your contribution!

mbolivar-nordic commented:
One more thing: the driver must at least be build-tested by modifying tests/drivers/build_all.

skullbox305 commented:
Thank you for the feedback! I made the necessary changes and also left some comments. I also added the build_all test for the sensor. 
For some reason the compliance checks fail due to a trailing semicolon warning in the driver initialization macro. I don`t know how to fix this. Can I ignore this?

mbolivar-nordic commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -0,0 +1,1002 @@
+		ch_supported = 2;
+	}
+
+	if (cfg->num_channels <= 0) {
```
Taking another look at this line, it's using `<= 0` on an unsigned type. That's a bit misleading, please use `== 0`.

skullbox305 replied:
Ok, now fixed

mbolivar-nordic reviewed (dismissed):
> For some reason the compliance checks fail due to a trailing semicolon warning in the driver initialization macro. I don`t know how to fix this. Can I ignore it?

Thanks for your update! I have a couple of follow ups but this looks basically OK to me now.

You should be able to fix the compliance issue by rebasing onto the latest master. The checkpatch warnings were recently turned into hard CI errors, so we've been working through disabling the ones that should be turned off. Trailing semicolons were turned off in 967aaee36100cb07d77a089ebeddb06d36049c55 which is in master now, so a rebase will fix your PR's CI for that.

mbolivar-nordic commented:
> One more thing: the driver must at least be build-tested by modifying tests/drivers/build_all.

Did you miss this one?

skullbox305 commented:
> 
> 
> > For some reason the compliance checks fail due to a trailing semicolon warning in the driver initialization macro. I don`t know how to fix this. Can I ignore it?
> 
> Thanks for your update! I have a couple of follow ups but this looks basically OK to me now.
> 
> You should be able to fix the compliance issue by rebasing onto the latest master. The checkpatch warnings were recently turned into hard CI errors, so we've been working through disabling the ones that should be turned off. Trailing semicolons were turned off in [967aaee](https://github.com/zephyrproject-rtos/zephyr/commit/967aaee36100cb07d77a089ebeddb06d36049c55) which is in master now, so a rebase will fix your PR's CI for that.

Thank you for the new feedback. I fixed all of your comments. Also the compliance check succeeded now. 
Looking forward to the feedback of the sensor maintainers for the other comments.

skullbox305 commented:
> 
> 
> > One more thing: the driver must at least be build-tested by modifying tests/drivers/build_all.
> 
> Did you miss this one?

No, I made the changes to `tests/drivers/build_all`. Did I forget to add something?

MaureenHelm reviewed (changes requested):
Overall really nice work, thank you for contributing!

mbolivar-nordic commented:
> No, I made the changes to `tests/drivers/build_all`. Did I forget to add something?

Sorry; I don't know how I missed that. Thanks.

skullbox305 commented:
> 
> 
> Overall really nice work, thank you for contributing!

Thank you!
I added the necessary conversion from raw to capacitance (pF) and frequency (MHz).

pabigot commented on include/drivers/sensor.h:
```diff
@@ -174,6 +174,24 @@ enum sensor_channel {
+	/** CH3 Capacitance, in Picofarad **/
+	SENSOR_CHAN_CAPACITANCE_CH3,
+
+	/** CH0 Frequency, in MHz **/
```
I'm not sure MHz is an ideal unit.  Hz would allow for frequencies like 0.5 Hz, i.e. every two seconds; or 0.000012 (a very inaccurate once-per-day).

This would limit the frequency to 2 GHz, though, which might be limiting in some applications.  Perhaps a future sensor API could select the best unit based on the value, and return either frequency in Hz or interval in some time unit, depending on which is more accurate.

This could be discussed in the API meeting.  Sticking with MHz or even kHz would be a reasonable resolution of such a discussion.

skullbox305 replied:
I think you're right. Hz would make more sense. I'm not sure if >2GHz would be necessary for sensors, but who knows. Would it be possible to have both units, Hz now and MHz later if needed? This way you could use Hz for <2GHz and MHz for values above. 
Also I agree with your other suggestion. Having the flexibility to get the value returned in the most accurate unit would be really nice.

I will wait for the resolution of the API meeting though and then change it based on that. Could you give me an update on the resolution?

henrikbrixandersen replied:
How about converting these sensor channel enums to be private to the FDC2X1X driver instead? Similar to how it is done for the MCUX ACMP sensor driver in https://github.com/zephyrproject-rtos/zephyr/blob/f71cd2aaf5e3f5942484a7ddb92e6192b2130dc6/include/drivers/sensor/mcux_acmp.h#L21?

carlescufi replied:
API meeting 26th Jan 2021:

Moving the enums to the driver-specific header is the concluded best approach for this.

skullbox305 replied:
Thank you both for your suggestion!  I moved the channel enums to the driver-specific header now

pabigot reviewed (commented):
Some thoughts on sensor units.  Note that we're now in 2.5 stabilization so this won't get merged until the 2.6 merge window re-opens.

mbolivar-nordic commented on drivers/sensor/fdc2x1x/fdc2x1x.c:
```diff
@@ -0,0 +1,1086 @@
+	};								   \
+									   \
+	static const struct fdc2x1x_config fdc2x1x_config_##n = {	   \
+		.i2c_name = DT_INST_BUS_LABEL(n),			   \
```
Hi again. I just see one more round of churn before this PR LGTM from a driver model perspective.

We are making changes to the way that sensor devices access their buses for Zephyr 2.6. These mean that getting them by name using device_get_binding is discouraged.

You should now store the device directly in the configuration structure in ROM by moving these fields to `struct fdc2x1x_config` from `struct fdc2x1x_data`:

```
	const struct device *bus;
	const struct device *gpio;
```

These should now be initialized using `DEVICE_DT_GET`. For example, `.bus = DEVICE_DT_GET(DT_INST_BUS(n))`.

You should then check the device in the init routine using `device_is_ready`.

Finally, fields like `i2c_name` which exist only to store devicetree label properties should be removed. You can use `bus->name` instead if you need a name.

For an example, see the drivers/led/led_pwm.c changes in https://github.com/zephyrproject-rtos/zephyr/pull/32558.

Thanks for your patience!

skullbox305 replied:
Hi. I addressed your comments. The bus and gpio devices are now initialized with `DEVICE_DT_GET` and checked  with `device_is_ready`. I also removed the `i2c_name` and both gpio labels.

skullbox305 committed:
samples: sensor: fdc2x1x: Add FDC2X1X sample application

This sample application periodically reads raw data from the FDC2X1X
sensor in polling mode or optionally with data ready trigger.
It is able to read the 12-Bit and 28-Bit, as well as the 2-Channel
and 4-Channel versions (FDC2112, FDC2114, FDC2212, FDC2214).
The 4-channel versions are chosen through devicetree properties.
The default is FDC2112.

Signed-off-by: Igor Knippenberg <igor.knippenberg@gmail.com>

skullbox305 committed:
drivers: sensors: fdc2x1x: Add driver for Texas Instruments FDC2X1X

Adds support for the Texas Instruments FDC2X1X Capacitance-to-Digital
Converter for Proximity and Level Sensing Applications.

Signed-off-by: Igor Knippenberg <igor.knippenberg@gmail.com>

mbolivar-nordic reviewed (approved):
Thanks! Sorry for the delay.

carlescufi commented:
@skullbox305 could you please rebase?

skullbox305 commented:
@mbolivar-nordic Not a problem. Thank you and the others for the review! It was really helpful
@carlescufi I rebased it now
